File: packages/agent-sdk/src/mcp-client.ts
/**
 * RenowideMcpClient — agent-as-client helper.
 *
 * Lets a deployed Renowide agent call back into the Renowide MCP server
 * with its own `rw_key_` token. The primary use case is cross-agent
 * collaboration:
 *
 *   - `delegate_task` to another hired specialist
 *   - `submit_proposal` for a governed action
 *   - `log_agent_action` / `write_guild_memory` to share context
 *   - `get_knowledge_base` / `get_attached_skills` to read hire-scoped config
 *
 * The Renowide MCP server speaks JSON-RPC 2.0 at
 *   POST https://renowide.com/api/v1/mcp
 * with `Authorization: Bearer <rw_key_...>`.
 *
 * This helper is intentionally thin — it wraps JSON-RPC framing and the
 * "result.content[0].text is JSON" convention so your handler gets a
 * plain object back.
 */
export interface RenowideMcpClientOptions {
  /**
   * Full MCP endpoint URL. Defaults to the value of `RW_MCP_URL`, then
   * to the production Renowide MCP URL. Override in tests.
   */
  mcpUrl?: string;
  /**
   * The `rw_key_...` token. Defaults to `RW_HIRE_TOKEN`
   * (the per-hire token Renowide injects into hosted agents), then to
   * `RENOWIDE_API_KEY` (your personal creator key).
   */
  token?: string;
  /** Request timeout in ms. Default 20_000. */
  timeoutMs?: number;
  /** Injected fetch implementation. Defaults to global fetch. */
  fetchImpl?: typeof fetch;
}

export interface McpToolInfo {
  name: string;
  description?: string;
  inputSchema?: unknown;
}

export class RenowideMcpClient {
  private readonly mcpUrl: string;
  private readonly token: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: RenowideMcpClientOptions = {}) {
    this.mcpUrl =
      options.mcpUrl ??
      process.env.RW_MCP_URL ??
      "https://renowide.com/api/v1/mcp";
    const token =
      options.token ??
      process.env.RW_HIRE_TOKEN ??
      process.env.RENOWIDE_API_KEY ??
      "";
    if (!token) {
      throw new Error(
        "RenowideMcpClient: missing token. Pass `token` or set RW_HIRE_TOKEN / RENOWIDE_API_KEY.",
      );
    }
    this.token = token;
    this.timeoutMs = options.timeoutMs ?? 20_000;
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  /** List every tool the Renowide MCP server exposes to this token. */
  async listTools(): Promise<McpToolInfo[]> {
    const res = await this.rpc<{ tools: McpToolInfo[] }>("tools/list", {});
    return res.tools ?? [];
  }

  /**
   * Call a Renowide MCP tool by name. Arguments are sent as-is under
   * `params.arguments`. The MCP server returns `result.content[]`;
   * when the first content item is JSON we parse it transparently, so
   * typical callers receive a plain object.
   */
  async callTool<T = unknown>(
    name: string,
    args: Record<string, unknown> = {},
  ): Promise<T> {
    const raw = await this.rpc<{ content?: Array<{ type: string; text: string }> }>(
      "tools/call",
      { name, arguments: args },
    );
    const first = raw?.content?.[0];
    if (first?.type === "text" && typeof first.text === "string") {
      const text = first.text.trim();
      if (text.startsWith("{") || text.startsWith("[")) {
        try {
          return JSON.parse(text) as T;
        } catch {
          // fall through; caller gets the raw text
        }
      }
      return first.text as unknown as T;
    }
    return raw as unknown as T;
  }

  private async rpc<T>(method: string, params: Record<string, unknown>): Promise<T> {
    const id = `rw_${Date.now().toString(36)}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(this.mcpUrl, {
        method: "POST",
        signal: controller.signal,
        headers: {
          "content-type": "application/json",
          accept: "application/json",
          authorization: `Bearer ${this.token}`,
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id,
          method,
          params,
        }),
      });
      const body: any = await response.json().catch(() => ({}));
      if (!response.ok || body?.error) {
        const message = body?.error?.message ?? `Renowide MCP ${method} failed (${response.status})`;
        const code = body?.error?.code ?? response.status;
        throw new RenowideMcpError(message, code, method);
      }
      return body.result as T;
    } finally {
      clearTimeout(timer);
    }
  }
}

export class RenowideMcpError extends Error {
  readonly code: number | string;
  readonly method: string;
  constructor(message: string, code: number | string, method: string) {
    super(message);
    this.name = "RenowideMcpError";
    this.code = code;
    this.method = method;
  }
}

File: python/renowide-agent-sdk/src/renowide_agent_sdk/mcp_client.py
"""
RenowideMcpClient — agent-as-client helper (Python).

Mirror of typescript/src/mcp-client.ts. Lets a deployed Renowide agent
call back into the Renowide MCP server with its own ``rw_key_`` token
to delegate tasks, submit proposals, log actions, read knowledge base,
etc.

The Renowide MCP server speaks JSON-RPC 2.0 at
    POST https://renowide.com/api/v1/mcp
with ``Authorization: Bearer <rw_key_...>``.
"""

from __future__ import annotations

import json
import os
from typing import Any, Dict, List, Optional

import httpx


DEFAULT_MCP_URL = "https://renowide.com/api/v1/mcp"


class RenowideMcpError(RuntimeError):
    """Raised when a Renowide MCP call fails."""

    def __init__(self, message: str, *, code: Any = None, method: str = "") -> None:
        super().__init__(message)
        self.code = code
        self.method = method


class RenowideMcpClient:
    """Thin client for the Renowide MCP server."""

    def __init__(
        self,
        *,
        mcp_url: Optional[str] = None,
        token: Optional[str] = None,
        timeout_s: float = 20.0,
        transport: Optional[httpx.BaseTransport] = None,
    ) -> None:
        self.mcp_url = (
            mcp_url
            or os.environ.get("RW_MCP_URL")
            or DEFAULT_MCP_URL
        )
        resolved_token = (
            token
            or os.environ.get("RW_HIRE_TOKEN")
            or os.environ.get("RENOWIDE_API_KEY")
            or ""
        )
        if not resolved_token:
            raise ValueError(
                "RenowideMcpClient: missing token. Pass `token` or set "
                "RW_HIRE_TOKEN / RENOWIDE_API_KEY."
            )
        self.token = resolved_token
        self.timeout_s = timeout_s
        self._client = httpx.Client(timeout=timeout_s, transport=transport)

    # ── Public API ────────────────────────────────────────────────────────

    def list_tools(self) -> List[Dict[str, Any]]:
        """Return every tool this token is allowed to call."""
        result = self._rpc("tools/list", {})
        return list(result.get("tools", []))

    def call_tool(self, name: str, arguments: Optional[Dict[str, Any]] = None) -> Any:
        """Call a Renowide MCP tool. JSON content is transparently decoded."""
        result = self._rpc("tools/call", {"name": name, "arguments": arguments or {}})
        content = result.get("content") or []
        if content:
            first = content[0]
            if first.get("type") == "text":
                text = (first.get("text") or "").strip()
                if text.startswith(("{", "[")):
                    try:
                        return json.loads(text)
                    except json.JSONDecodeError:
                        return first["text"]
                return first.get("text")
        return result

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "RenowideMcpClient":
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()

    # ── Internals ─────────────────────────────────────────────────────────

    def _rpc(self, method: str, params: Dict[str, Any]) -> Dict[str, Any]:
        payload = {
            "jsonrpc": "2.0",
            "id": f"rw_{os.urandom(4).hex()}",
            "method": method,
            "params": params,
        }
        try:
            response = self._client.post(
                self.mcp_url,
                json=payload,
                headers={
                    "authorization": f"Bearer {self.token}",
                    "accept": "application/json",
                },
            )
        except httpx.HTTPError as exc:
            raise RenowideMcpError(
                f"Transport error on {method}: {exc}",
                code="transport_error",
                method=method,
            ) from exc

        try:
            body = response.json()
        except ValueError:
            body = {}

        if response.status_code >= 400 or body.get("error"):
            err = body.get("error") or {}
            raise RenowideMcpError(
                err.get("message") or f"Renowide MCP {method} failed ({response.status_code})",
                code=err.get("code", response.status_code),
                method=method,
            )
        return body.get("result") or {}
